### PR TITLE
[FLINK-12424] [table-planner-blink] Supports dag (multiple-sinks query) optimization

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Supports dag (multiple-sinks query) optimization*

## Brief change log

  - *Add execute method in TableEnvironment to compile the multiple-sinks plan to transformations and execute the job*
  - *Add implementation of explain method to display the multiple-sinks plan*
  - *Add RelNodeBlock to decompose the multiple-sinks plan to different block*
  - *Update Optimizer implementation to handle multiple-sinks optimization*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added ExplainTest to validate the explain result for different kind of queries*
  - *Added DagOptimizationTest to validate the optimized plan for different kind of queries*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/optimize/OptimizerBase.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.calcite.rel.{RelNode, RelShuttleImpl}
+
+import java.util
+
+import scala.collection.JavaConversions._
+
+/**
+  * Base class for [[Optimizer]].
+  */
+abstract class OptimizerBase extends Optimizer {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/optimize/OptimizerBase.scala]: I'm not sure this is some kind of base optimizer for all the optimizers. It provides some specific functionality with `IntermediateRelNodeTable`, and not every optimizer should have this concept. Maybe you can change the name to a more dedicated form.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecIntermediateTableScan.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.flink.table.plan.nodes.common.CommonIntermediateTableScan
+
+import org.apache.calcite.plan._
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.`type`.RelDataType
+
+/**
+  * Batch physical TableScan that wraps [[org.apache.flink.table.plan.schema.IntermediateRelTable]]
+  */
+class BatchExecIntermediateTableScan(
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecIntermediateTableScan.scala]: A `IntermediateRelTable` needs to be wrap in a `TableScan`, so those `TableScan`s represent table scan `RelNode` for different optimization phases. `IntermediateRelTable` will be expanded to real `RelNode` sub-tree after Calcite optimization, so those `TableScan`s are not `ExecNode` and do not need translate to plan.

[the remark this answers is not in the document]
Comment (KurtYoung) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecIntermediateTableScan.scala]: Does a common physical table scan operator enough for this case? There is no need to have two concrete `BatchExecIntermediateTableScan` and `StreamExecIntermediateTableScan`

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/optimize/OptimizerBase.scala (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/optimize/OptimizerBase.scala]: yes, i think `CommonSubGraphBasedOptimizer` is better.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecIntermediateTableScan.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecIntermediateTableScan.scala]: we must have them both, `StreamExecIntermediateTableScan` needs to implement retraction methods (like `isAccRetract`, `producesUpdates`) which return the corresponding properties from the wrapped `RelNode` table.

**Pull request merged**